File: API/src/main/java/com/ltizzi/ecommerce/model/user/UserMapper.java
package com.ltizzi.ecommerce.model.user;

import com.ltizzi.ecommerce.model.cart.CartMapper;
import com.ltizzi.ecommerce.model.cart.CartResponse;
import com.ltizzi.ecommerce.model.purchase.PurchaseMapper;
import com.ltizzi.ecommerce.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.client.HttpClientErrorException;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Leonardo Terlizzi
 */

@Component
public class UserMapper {

    @Autowired
    private CartMapper cartMapper;

    @Autowired
    private PurchaseMapper purchaseMapper;

    @Autowired
    private UserRepository userRepo;

    public UserResponse toUserResponse(UserEntity user) {
        UserResponse userRes = new UserResponse();
        userRes.setUser_id(user.getUser_id());
        userRes.setUsername(user.getUsername());
        userRes.setName(user.getName());
        userRes.setLastname(user.getLastname());
        userRes.setEmail(user.getEmail());
        List<CartResponse> carts = cartMapper.toArrayCartResponse(user.getCarts());
        userRes.setCarts(carts);
        userRes.setAvatar(user.getAvatar());
        userRes.setBirthday(user.getBirthday());
        userRes.setPurchases(purchaseMapper.toArrayPurchaseResponse(user.getPurchases()));

        return userRes;
    }

    public UserEntity toUserEntity(UserRequest userReq) throws HttpClientErrorException.NotFound {
        UserEntity user = new UserEntity();
        if (userReq.getUser_id() != null) {
            user = userRepo.findById(userReq.getUser_id()).orElseThrow();
        }
        user.setAvatar(userReq.getAvatar());
        user.setBirthday(userReq.getBirthday());
        user.setCarts(cartMapper.toArrayCartEntity(userReq.getCarts()));
        user.setEmail(userReq.getEmail());
        user.setName(userReq.getName());
        user.setLastname(userReq.getLastname());
        user.setUsername(userReq.getUsername());
        user.setPurchases(purchaseMapper.toArrayPurchaseEntity(userReq.getPurchases()));

        return user;
    }

    public List<UserResponse> toArrayUserResponse(List<UserEntity> users) {
        List<UserResponse> usersReq = new ArrayList<>();
        users.forEach(user -> {
            usersReq.add(toUserResponse(user));
        });
        return usersReq;
    }
}

File: Server/src/dto/mappers/user.mapper.ts
import { User } from "../../models/User";
import { UserService } from "../../services/UserService";
import { CartRequest } from "../requests/cart.request";
import { PurchaseRequest } from "../requests/purchase.request";
import { UserRequest } from "../requests/user.request";
import { CartResponse } from "../responses/cart.response";
import { PurchaseResponse } from "../responses/purchase.response";
import { UserResponse } from "../responses/user.response";
import { CartMapper } from "./cart.mapper";
import { PurchaseMapper } from "./purchase.mapper";

const userServ = new UserService();
const cartMapper = new CartMapper();
const purchMapper = new PurchaseMapper();

export class UserMapper {
  toUserResponse(user: User): UserResponse {
    const res = {} as UserResponse;
    res.user_id = user.user_id;
    res.username = user.username;
    res.name = user.name;
    res.lastname = user.lastname;
    res.email = user.email;
    res.avatar = user.avatar;
    res.birthday = user.birthday;
    if (user.carts) res.carts = cartMapper.toArrayCartResponse(user.carts);
    if (user.purchases)
      res.purchases = purchMapper.toArrayPurchaseResponse(user.purchases);
    res.roles = user.roles;
    return res;
  }

  async toUserEntity(userReq: UserRequest | UserResponse): Promise<User> {
    let user: User = {} as User;
    if (userReq.user_id) {
      user = (await userServ.getUserById(userReq.user_id)) as User;
    }
    user.avatar = userReq.avatar;
    user.birthday = userReq.birthday;
    if (userReq.carts) {
      user.carts = await cartMapper.toArrayCartEntity(userReq.carts);
    }
    user.email = userReq.email;
    user.name = userReq.name;
    user.lastname = userReq.lastname;
    user.username = userReq.username;
    if (userReq.purchases) {
      user.purchases = await purchMapper.toArrayPurchaseEntity(
        userReq.purchases
      );
    }
    if (userReq.roles) user.roles = userReq.roles;
    return user;
  }

  toArrayUserResponse(users: Array<User>): Array<UserResponse> {
    let usersRes: Array<UserResponse> = [];
    users.forEach(async (user) => usersRes.push(this.toUserResponse(user)));
    return usersRes;
  }
}
